### DISCUSSION TITLE
Add a method in setRawOriginal() in Illuminate/Database/Eloquent/Concerns/HasAttributes

### DISCUSSION THREAD

**Discussion by Ronan-Lenor** [laravel/framework#46450] [Ideas] [answered]
Hello,

at a certain point in my project, i needed to do this :
```php
    // allow only one address to be a default billing address
    if (array_key_exists('is_billing', $validated))
    {
        // 1. reset all addresses that could be set as default billing address (updating here won't affect the current model $address)
        Auth::user()->addresses()->update(['is_billing' => 0]);

        // 2. if the given $address model already have an "original" property : is_billing = 1
        //    (case : the user submit the <form> without changing something in the default billing address)
        //    then force the "original" property value to 0 to trigger the update of the field in the database
        //    (\Illuminate\Database\Eloquent\Concerns\HasAttributes > isDirty/hasChanges)
        $address->is_billing = 0;
        $address->syncOriginalAttribute('is_billing');

        // 3. reset the $address model "attributes" property to 1 before saving.
        $address->is_billing = 1;
    }
```

and i was wondering if you could add a `setRawOriginal(...)` method in `/Illuminate/Database/Eloquent/Concerns/HasAttributes.php`
so we could do the same than above in only one line like so:
```php
    $address->setRawOriginal('is_billing', 0);
```

Thanks :P

**Comment by henzeb** [comment, marked as answer]
We have setRawAttributes, does that do what you want to achieve?

**Reply by Ronan-Lenor** [reply]
Oh yeah, it is true that there is that method. But what i noticed when i used it, is that if i only give `['attribute' => true]` in the parameters, it will delete every other attributes.

Actually, i would love to change only 1 attribute of the model, without redefining all the others :P

(I have edited my first post to make it look like more what i'm expecting)

**Reply by bert-w** [reply]
https://github.com/laravel/framework/pull/30885

The PR for `setRawAttribute()` got reverted because it clashes with calling `$myModel->raw = 'abc'` (which calls the set-mutator). Apparently having a `raw` column is pretty common.

You could create a PR for something like `public function setAttribute($key, $value, $raw = false)` to avoid such nameclash, but this adds another argument to an existing function.

**Reply by Ronan-Lenor** [reply]
Actually, unlike `setRawAttribute()` from https://github.com/laravel/framework/pull/30885 that modify the `protected $attributes`, the purpose of a method `setRawOriginal()` is to modifying only one value stored in the `protected $original` property. (So Mutators are not concerned)

i should had give a better example on my first post (that indeed modify the attributes & original properties) where in fact, i only wanted the `orginal` property to be changed. i will edit my first post to make it more obvious.

**Reply by bert-w** [reply]
So why exactly can't you do `$address->update(['is_billing' => 1]);` or something like 
```php
$address->is_billing = 1;
// ... more logic
$address->save();
```

**Reply by Ronan-Lenor** [reply]
Because, as i writed in the parenthesis `(case : the user submit the <form> without changing something in the default billing address)`, the `$address` model has already the property `is_billing` set to `1`.

So, as the `$address`  didn't change, doing `$address->save()` won't trigger any database query update.

And as i use the below script help to cover multiple cases (such as `Create` (where the `$address->id` don't exist))
`Auth::user()->addresses()->update(['is_billing' => 0]);`
Then, the databases entry that is targeted by `$address` get also affected and is set to `0`.
Resulting by storing in the database `0` when `1` is expected.

and that's why i have to indicate to the model `$address` that database changes have been made, concerning this model, by changing the property `is_billing` stored in the `original` property.

Now that i think about it, there's another solution that imply to do something like so:
```php
    $query = Auth::user()->addresses();

    if ($address->id != null)
    {
        $query->whereNot('id', $address->id);
    }

    $query->update(['is_billing' => 0]);
```
at least i won 1 query in less to the database.

Nonetheless, regarding of how the method  `syncOriginalAttribute()` work, wouldn't it be it great to have a `setRawOriginal()` that would do the same job but for only one property?

**Reply by bert-w** [reply]
Yea I think that is a valid solution:
```php
// Reset all billing addresses except the current reference.
Auth::user()->addresses()->when($address->id)->whereNot('id', $address->id)->update(['is_billing' => 0]);
// Update the current reference.
$address->update(['is_billing' => 1]);
```
Reset all addresses to 0 for which you don't have a local object reference. Then do whatever you need to do for which you *do* have a reference (like `$address`). This way the state of the objects is never messed up.
